lib/DataFileUtil/DataFileUtilImpl.py

```diff
@@ -245,6 +251,271 @@ def make_handle(self, shock_data, token):
+        copy_file_path = os.path.join(self.tmp, file_name)
+
+        return copy_file_path
+
+    def _download_to_file(self, file_url, copy_file_path):
+        """
+        _download_to_file: download url content to file
+
+        params:
+        file_url: direct download URL
+        copy_file_path: output file saving path
+
+        """
+
+        self.log('Connecting and downloading web source: {}'.format(
+                                                                file_url))
+        try:
+            online_file = urllib2.urlopen(file_url)
+        except urllib2.HTTPError as e:
+            raise ValueError(
+                "The server error\nURL: {}\nError code: {}".format(
+                                                        file_url, e.code))
+        except urllib2.URLError as e:
+            raise ValueError(
+                "Failed to reach URL: {}\nReason: {}".format(
+                                                      file_url, e.reason))
+        else:
+            total_size = int(online_file.info().getheader('Content-Length').strip())
+            CHUNK = 128 * 1024 * 1024
+            downloaded = 0
+            with closing(online_file):
+                with open(copy_file_path, 'wb') as output:
+                    start_time = time.time()
+                    while True:
+                      chunk = online_file.read(CHUNK)
+                      if not chunk: break
+                      output.write(chunk)
+                      downloaded += len(chunk)
+                      process = float(downloaded) / total_size * 100
+                      used_time = time.time() - start_time
+                      self.log('downloaded: {:.2f}%, '.format(process) + 
+                                        'used: {:.2f}s'.format(used_time))
+
+            self.log('Downloaded file to {}'.format(copy_file_path))
+
+    def _download_direct_download_link(self, file_url):
+        """
+        _download_direct_download_link: direct download link handler
+
+        params:
+        file_url: direct download URL
+        copy_file_path: output file saving path
+
+        """
+
+        copy_file_path = self._retrieve_filepath(file_url, 'Direct Download')
+        self._download_to_file(
+            file_url, copy_file_path)
+
+        copy_file_path = self._unpack(copy_file_path, True)
+
+        return copy_file_path
+
+    def _download_dropbox_link(self, file_url):
+        """
+        _download_dropbox_link: dropbox download link handler
+                                file needs to be shared publicly
+
+        params:
+        file_url: dropbox download link
+
+        """
+        if not file_url.startswith('https://www.dropbox.com/'):
+            raise ValueError('Invalid DropBox Link: {}'.format(file_url))
+
+        self.log('Connecting DropBox link: {}'.format(file_url))
+        # translate dropbox URL for direct download
+        if "?" not in file_url:
```

Comment:
I don't see a test for this case - I could've sworn we added one originally as part of the RU review.